File: Semester-4/ProgLanguages-CourseProject-DB/py_sources/fill.py
# -*- coding: utf-8 -*-

# Подключение модулей
import main
import const
import database
import out


# Класс исключения из-за неправильного пола
class GenderError(Exception):
    def __init__(self, value):
        self.value = value
    def __str__(self):
        return repr(self.value)


# Класс исключения из-за неправильного кода
class BadIdError(Exception):
    def __init__(self, value):
        self.value = value
    def __str__(self):
        return repr(self.value)


def check_date(date):
    if date.day < 0 or date.day > 31 or date.month < 0 or date.month > 12 or date.year < 0:
        print("ПРЕДУПРЕЖДЕНИЕ: Скорее всего, введена не дата")


def input_gender(desc):
    temp = input(desc)
    if temp[0] == "m" or temp[0] == "M":
        return database.Gender.male
    elif temp[0] == "f" or temp[0] == "F":
        return database.Gender.female
    else:
        raise GenderError


def input_date(date, desc):
    print(desc)
    date.day = input("\tВведите день -> ")
    date.month = input("\tВведите месяц -> ")
    date.year = input("\tВведите год -> ")


def fill_employee_table(row):
    database.employee_table.append(database.EmployeeT())
    main.activated = 1

    try:
        database.employee_table[row].employee_id = int(input("Введите код сотрудника (больше нуля) -> "))
        if database.employee_table[row].employee_id <= 0:
            raise BadIdError
        database.employee_table[row].name = input("Введите ФИО -> ")
        database.employee_table[row].age = int(input("Введите возраст -> "))
        if database.employee_table[row].age < 0:
            print("ПРЕДУПРЕЖДЕНИЕ: Скорее всего, введен не возраст")
        database.employee_table[row].gender = input_gender("Введите пол -> ")
        database.employee_table[row].address = input("Введите адрес -> ")
        database.employee_table[row].phone = input("Введите телефон -> ")
        database.employee_table[row].passport = input("Введите паспортные данные -> ")
        out.print_position_info()
        database.employee_table[row].position_id = int(input("Введите код должности (больше нуля) -> "))
        if database.employee_table[row].position_id < 0:
            raise BadIdError
        main.table_row[const.EMPLOYEE_TABLE] += 1
    except ValueError:
        database.employee_table.pop()
        print("ОШИБКА: не удалось распознать в строке целочисленное значение")
    except GenderError:
        database.employee_table.pop()
        print("ОШИБКА: не удалось распознать пол")
    except BadIdError:
        database.employee_table.pop()
        print("ОШИБКА: введен код меньше нуля, таблица не добавится")


def fill_position_table(row):
    database.position_table.append(database.PositionT())
    main.activated = 1

    try:
        database.position_table[row].position_id = int(input("Введите код должности (больше нуля) -> "))
        if database.position_table[row].position_id <= 0:
            raise BadIdError
        database.position_table[row].name = input("Введите наименование -> ")
        database.position_table[row].salary = int(input("Введите оклад -> "))
        database.position_table[row].duties = input("Введите обязанности -> ")
        database.position_table[row].requirements = input("Введите требования -> ")
        main.table_row[const.POSITION_TABLE] += 1
    except ValueError:
        database.position_table.pop()
        print("ОШИБКА: не удалось распознать в строке целочисленное значение")


def fill_brand_table(row):
    database.brand_table.append(database.BrandT())
    main.activated = 1

    try:
        database.brand_table[row].brand_id = int(input("Введите код марки (больше нуля) -> "))
        if database.brand_table[row].brand_id <= 0:
            raise BadIdError
        database.brand_table[row].name = input("Введите наименование -> ")
        database.brand_table[row].specs = input("Введите технические характеристики -> ")
        database.brand_table[row].description = input("Введите описание -> ")
        main.table_row[const.BRAND_TABLE] += 1
    except ValueError:
        database.brand_table.pop()
        print("ОШИБКА: не удалось распознать в строке целочисленное значение")


def fill_service_table(row):
    database.service_table.append(database.ServiceT())
    main.activated = 1
    try:
        database.service_table[row].service_id = int(input("Введите код услуги (больше нуля) -> "))
        if database.service_table[row].service_id <= 0:
            raise BadIdError
        database.service_table[row].name = input("Введите наименование -> ")
        database.service_table[row].description = input("Введите описание -> ")
        database.service_table[row].price = int(input("Введите цену -> "))
        main.table_row[const.SERVICE_TABLE] += 1
    except ValueError:
        database.brand_table.pop()
        print("ОШИБКА: не удалось распознать в строке целочисленное значение")


def fill_car_table(row):
    database.car_table.append(database.CarT())
    main.activated = 1

    try:
        database.car_table[row].car_id = int(input("Введите код автомобиля (больше нуля) -> "))
        if database.car_table[row].car_id <= 0:
            raise BadIdError
        out.print_brand_info()
        database.car_table[row].brand_id = int(input("Введите код марки (больше нуля) -> "))
        if database.car_table[row].brand_id <= 0:
            raise BadIdError
        out.print_employee_info()
        database.car_table[row].employee_id = int(input("Введите код сотрудникамеханика (больше нуля) -> "))
        if database.car_table[row].employee_id <= 0:
            raise BadIdError
        database.car_table[row].reg_number = int(input("Введите регистрационный номер -> "))
        database.car_table[row].body_number = int(input("Введите номер кузова -> "))
        database.car_table[row].engine_number = int(input("Введите номер двигателя -> "))
        database.car_table[row].release_year = int(input("Введите год выпуска -> "))
        database.car_table[row].mileage = int(input("Введите пробег -> "))
        database.car_table[row].price = int(input("Введите цену автомобиля -> "))
        database.car_table[row].rent_price = int(input("Введите цену дня проката -> "))
        input_date(database.car_table[row].maintenance_date, "Введите дату последнего ТО")
        check_date(database.car_table[row].maintenance_date)
        database.car_table[row].special_marks = input("Введите специальные отметки -> ")
        database.car_table[row].return_mark = input("Введите отметку о возврате ->")
        main.table_row[const.CAR_TABLE] += 1
    except ValueError:
        database.car_table.pop()
        print("ОШИБКА: не удалось распознать в строке целочисленное значение")


def fill_client_table(row):
    database.client_table.append(database.ClientT())
    main.activated = 1

    try:
        database.client_table[row].client_id = int(input("Введите код клиента (больше нуля) -> "))
        if database.client_table[row].client_id <= 0:
            raise BadIdError
        database.client_table[row].name = input("Введите ФИО -> ")
        database.client_table[row].gender = input_gender("Введите пол -> ")
        input_date(database.client_table[row].birth_date, "Введите дату рождения")
        check_date(database.client_table[row].birth_date)
        database.client_table[row].address = input("Введите адрес -> ")
        database.client_table[row].phone = input("Введите телефон -> ")
        database.client_table[row].passport = input("Введите паспортные данные -> ")
        main.table_row[const.CLIENT_TABLE] += 1
    except ValueError:
        database.client_table.pop()
        print("ОШИБКА: не удалось распознать в строке целочисленное значение")
    except GenderError as error:
        database.client_table.pop()
        print(error)


def fill_rent_table(row):
    database.rent_table.append(database.RentT())
    main.activated = 1

    try:
        out.print_car_info()
        database.rent_table[row].car_id = int(input("Введите код автомобиля (больше нуля) -> "))
        if database.rent_table[row].car_id <= 0:
            raise BadIdError
        out.print_client_info()
        database.rent_table[row].client_id = int(input("Введите код клиента (больше нуля) -> "))
        if database.rent_table[row].client_id <= 0:
            raise BadIdError
        out.print_employee_info()
        database.rent_table[row].employee_id = int(input("Введите код сотрудника (больше нуля) -> "))
        if database.rent_table[row].employee_id <= 0:
            raise BadIdError
        out.print_service_info()
        database.rent_table[row].service1_id = int(input("Введите код услуги 1 (больше нуля) -> "))
        if database.rent_table[row].service1_id <= 0:
            raise BadIdError
        database.rent_table[row].service2_id = int(input("Введите код услуги 2 (больше нуля) -> "))
        if database.rent_table[row].service2_id <= 0:
            raise BadIdError
        database.rent_table[row].service3_id = int(input("Введите код услуги 3 (больше нуля) -> "))
        if database.rent_table[row].service3_id <= 0:
            raise BadIdError
        input_date(database.rent_table[row].delivery_date, "Введите дату выдачи")
        check_date(database.rent_table[row].delivery_date)
        database.rent_table[row].rent_time = int(input("Введите срок проката -> "))
        input_date(database.rent_table[row].return_date, "Введите дату возврата")
        check_date(database.rent_table[row].return_date)
        database.rent_table[row].rent_price = int(input("Введите цену проката -> "))
        database.rent_table[row].payment_mark = input("Введите отметку об оплате -> ")
        main.table_row[const.RENT_TABLE] += 1
    except ValueError:
        database.client_table.pop()
        print("ОШИБКА: не удалось распознать в строке целочисленное значение")

File: Semester-4/ProgLanguages-CourseProject-DB/c_sources/fill.c
#include "database.h"

int input_string(char *what, const int buffer, char *description)
{
	printf("\t\t%s", description);
	if( !fgets(what, buffer, stdin) )
	{
		fprintf(stderr, "ОШИБКА: Невозможно прочитать строку\n");
		return 1;
	}	
	what[strlen(what) - 1] = '\0';

	fflush(stdin);
	return 0;
}

int input_int(int *what, const int buffer, char *description)
{
	char *temp = (char *) malloc(buffer * sizeof(char));

	printf("\t\t%s", description);
	if( !fgets(temp, buffer, stdin) )
	{
		fprintf(stderr, "ОШИБКА: Невозможно прочитать строку\n");
		return 1;
	}	
	if( !sscanf(temp, "%d", what) )
	{
		fprintf(stderr, "ОШИБКА: Невозможно распознать в строке число\n");
		return 1;
	}

	fflush(stdin);
	free(temp);
	return 0;
}

int input_gender(enum gender_t *what, const int buffer, char *description)
{
	char *temp = (char *) malloc(buffer * sizeof(char));

	printf("\t\t%s", description);
	if( !fgets(temp, buffer, stdin) )
	{
		fprintf(stderr, "ОШИБКА: Невозможно прочитать строку\n");
		return 1;
	}	
	if( temp[0] == 'м' || temp[0] == 'М')
		*what = male;
	else if( temp[0] == 'ж' || temp[0] == 'Ж')
		*what = female;
	else
	{
		fprintf(stderr, "ОШИБКА: Не удалось распознать пол\n");
		return 1;
	}

	fflush(stdin);
	free(temp);
	return 0;
}

int input_date(date *what, const int buffer, char *description)
{
	printf("\t\t%s\n", description);

	if( input_int(&what->day, buffer,
		"\tВведите день  -> ") )
		return 1;
	if( input_int(&what->month, buffer,
		"\tВведите месяц -> ") )
		return 1;
	if( input_int(&what->year, buffer,
		"\tВведите год   -> ") )
		return 1;

	return 0;
}

int fill_employee_table(int row)
{
	activated = 1;
	employee_table = realloc( employee_table, sizeof(employee_struct)*(table_row[EMPLOYEE_TABLE]+1) );
	
	printf("\t\tЗаполните таблицу \"Сотрудники\"\n\n");

	if( input_int( &employee_table[row].employee_id, 10,
		"Введите код сотрудника (больше нуля) -> ") )
		return 1;
	if( employee_table[row].employee_id <= 0 )
	{
		fprintf(stderr, "ОШИБКА: введен код меньше нуля, таблица не добавится");
		return 1;
	}

	if( input_string( employee_table[row].name, 255,
		"Введите ФИО -> ") )
		return 1;

	if( input_int( &employee_table[row].age, 4,
		"Введите возраст -> ") )
		return 1;
	if( (employee_table[row].age < 0) || (employee_table[row].age > 110) )
		fprintf(stderr, "ПРЕДУПРЕЖДЕНИЕ: Скорее всего, введен не возраст\n");

	if( input_gender( &employee_table[row].gender, 10,
		"Введите пол -> ") )
		return 1;

	if( input_string( employee_table[row].address, 255,
		"Введите адрес -> ") )
		return 1;

	if( input_string( employee_table[row].phone, 20,
		"Введите телефон -> ") )
		return 1;
	if( !isdigit(employee_table[row].phone[2]) )
		fprintf(stderr, "ПРЕДУПРЕЖДЕНИЕ: Скорее всего, введен не телефон\n");

	if( input_string( employee_table[row].passport, 20,
		"Введите паспортные данные -> ") )
		return 1;
	if( !isdigit(employee_table[row].passport[1]) )
		fprintf(stderr, "ПРЕДУПРЕЖДЕНИЕ: Скорее всего, введены не паспортные данные\n");

	print_position_info();
	if( input_int( &employee_table[row].position_id, 10,
		"Введите код должности (больше нуля) -> ") )
		return 1;
	if( employee_table[row].position_id <= 0 )
	{
		fprintf(stderr, "ОШИБКА: введен код меньше нуля, таблица не добавится");
		return 1;
	}

	table_row[EMPLOYEE_TABLE]++;
	return 0;
}

int fill_position_table(int row)
{
	activated = 1;
	position_table = realloc( position_table, sizeof(position_struct)*(table_row[POSITION_TABLE]+1) );

	printf("\t\tЗаполните таблицу \"Должности\"\n\n");
	if( input_int(&position_table[row].position_id, 10,
		"Введите код должности (больше нуля) -> ") )
		return 1;
	if( position_table[row].position_id <= 0 )
	{
		fprintf(stderr, "ОШИБКА: введен код меньше нуля, таблица не добавится");
		return 1;
	}

	if( input_string(position_table[row].name, 255,
		"Введите наименование -> ") )
		return 1;

	if( input_int(&position_table[row].salary, 10,
		"Введите оклад -> ") )
		return 1;
	if( position_table[row].salary < 0 )
		fprintf(stderr, "ПРЕДУПРЕЖДЕНИЕ: Скорее всего, введен не оклад\n");

	if( input_string(position_table[row].duties, 255,
		"Введите обязанности -> ") )
		return 1;

	if( input_string(position_table[row].requirements, 255,
		"Введите требования -> ") )
		return 1;

	table_row[POSITION_TABLE]++;
	return 0;
}

int fill_brand_table(int row)
{
	activated = 1;
	brand_table = realloc( brand_table, sizeof(brand_struct)*(table_row[BRAND_TABLE]+1) );

	printf("\t\tЗаполните таблицу \"Марки автомобилей\"\n\n");
	if( input_int(&brand_table[row].brand_id, 10,
		"Введите код марки (больше нуля) -> ") )
		return 1;
	if( brand_table[row].brand_id <= 0 )
	{
		fprintf(stderr, "ОШИБКА: введен код меньше нуля, таблица не добавится");
		return 1;
	}

	if( input_string(brand_table[row].name, 255,
		"Введите наименование -> ") )
		return 1;

	if( input_string(brand_table[row].specs, 255,
		"Введите технические характеристики -> ") )
		return 1;

	if( input_string(brand_table[row].description, 255,
		"Введите описание -> ") )
		return 1;

	table_row[BRAND_TABLE]++;
	return 0;
}

int fill_service_table(int row)
{
	activated = 1;
	service_table = realloc( service_table, sizeof(service_struct)*(table_row[SERVICE_TABLE]+1) );

	printf("\t\tЗаполните таблицу \"Дополнительные услуги\"\n\n");
	if( input_int(&service_table[row].service_id, 10,
		"Введите код услуги (больше нуля) -> ") )
		return 1;
	if( service_table[row].service_id <= 0 )
	{
		fprintf(stderr, "ОШИБКА: введен код меньше нуля, таблица не добавится");
		return 1;
	}

	if( input_string(service_table[row].name, 255,
		"Введите наименование -> ") )
		return 1;

	if( input_string(service_table[row].description, 255,
		"Введите описание -> ") )
		return 1;

	if( input_int(&service_table[row].price, 10,
		"Введите цену -> ") )
		return 1;
	if( service_table[row].price < 0 )
		fprintf(stderr, "ПРЕДУПРЕЖДЕНИЕ: Скорее всего, введена не цена\n");

	table_row[SERVICE_TABLE]++;
	return 0;
}

int fill_car_table(int row)
{
	activated = 1;
	car_table = realloc( car_table, sizeof(car_struct)*(table_row[CAR_TABLE]+1) );

	printf("\t\tЗаполните таблицу \"Автомобили\"\n\n");
	if( input_int(&car_table[row].car_id, 10,
		"Введите код автомобиля (больше нуля) -> ") )
		return 1;
	if( car_table[row].car_id <= 0 )
	{
		fprintf(stderr, "ОШИБКА: введен код меньше нуля, таблица не добавится");
		return 1;
	}

	print_brand_info();
	if( input_int(&car_table[row].brand_id, 10,
		"Введите код марки -> ") )
		return 1;

	print_position_info();
	if( input_int(&car_table[row].employee_id, 10,
		"Введите код сотрудника-механика -> ") )
		return 1;

	if( input_int(&car_table[row].reg_number, 20,
		"Введите регистрационный номер -> ") )
		return 1;

	if( input_int(&car_table[row].body_number, 20,
		"Введите номер кузова -> ") )
		return 1;

	if( input_int(&car_table[row].engine_number, 20,
		"Введите номер двигателя -> ") )
		return 1;

	if( input_int(&car_table[row].release_year, 5,
		"Введите год выпуска -> ") )
		return 1;
	if( car_table[row].release_year < 0 )
		fprintf(stderr, "ПРЕДУПРЕЖДЕНИЕ: Скорее всего, введен не год\n");

	if( input_int(&car_table[row].mileage, 10,
		"Введите пробег -> ") )
		return 1;
	if( car_table[row].mileage < 0 )
		fprintf(stderr, "ПРЕДУПРЕЖДЕНИЕ: Скорее всего, введен не пробег\n");

	if( input_int(&car_table[row].price, 10,
		"Введите цену автомобиля -> ") )
		return 1;
	if( car_table[row].price < 0 )
		fprintf(stderr, "ПРЕДУПРЕЖДЕНИЕ: Скорее всего, введена не цена\n");

	if( input_int(&car_table[row].rent_price, 20,
		"Введите цену дня проката -> ") )
		return 1;
	if( car_table[row].rent_price < 0 )
		fprintf(stderr, "ПРЕДУПРЕЖДЕНИЕ: Скорее всего, введена не цена проката\n");

	if( input_date(&car_table[row].maintenance_date, 10,
		"Введите дату последнего ТО") )
		return 1;
	if( car_table[row].maintenance_date.day < 0 ||
		car_table[row].maintenance_date.day > 31 ||
		car_table[row].maintenance_date.month < 0 ||
		car_table[row].maintenance_date.month > 12 ||
		car_table[row].maintenance_date.year < 0 )
		fprintf(stderr, "ПРЕДУПРЕЖДЕНИЕ: Скорее всего, введена не дата последнего ТО\n");

	if( input_string(car_table[row].special_marks, 255,
		"Введите специальные отметки -> ") )
		return 1;

	if( input_string(car_table[row].return_mark, 255,
		"Введите отметку о возврате -> ") )
		return 1;

	table_row[CAR_TABLE]++;
	return 0;
}

int fill_client_table(int row)
{
	activated = 1;
	client_table = realloc( client_table, sizeof(client_struct)*(table_row[CLIENT_TABLE]+1) );

	printf("\t\tЗаполните таблицу \"Клиенты\"\n\n");
	if( input_int(&client_table[row].client_id, 10,
		"Введите код клиента (больше нуля) -> ") )
		return 1;
	if( client_table[row].client_id <= 0 )
	{
		fprintf(stderr, "ОШИБКА: введен код меньше нуля, таблица не добавится");
		return 1;
	}

	if( input_string(client_table[row].name, 255,
		"Введите ФИО -> ") )
		return 1;

	if( input_gender(&client_table[row].gender, 10,
		"Введите пол -> ") )
		return 1;

	if( input_date(&client_table[row].birth_date, 10,
		"Введите дату рождения") )
		return 1;
	if( client_table[row].birth_date.day < 0 ||
		client_table[row].birth_date.day > 31 ||
		client_table[row].birth_date.month < 0 ||
		client_table[row].birth_date.month > 12 ||
		client_table[row].birth_date.year < 0 )
		fprintf(stderr, "ПРЕДУПРЕЖДЕНИЕ: Скорее всего, введена не дата рождения\n");

	if( input_string(client_table[row].address, 255,
		"Введите адрес -> ") )
		return 1;

	if( input_string(client_table[row].phone, 20,
		"Введите телефон -> ") )
		return 1;
	if( !isdigit(client_table[row].phone[2]) )
		fprintf(stderr, "ПРЕДУПРЕЖДЕНИЕ: Скорее всего, введен не телефон\n");

	if( input_string(client_table[row].passport, 20,
		"Введите паспортные данные -> ") )
		return 1;
	if( !isdigit(client_table[row].passport[0]) )
		fprintf(stderr, "ПРЕДУПРЕЖДЕНИЕ: Скорее всего, введены не паспортные данные\n");

	table_row[CLIENT_TABLE]++;
	return 0;
}

int fill_rent_table(int row)
{
	activated = 1;
	rent_table = realloc( rent_table, sizeof(rent_struct)*(table_row[RENT_TABLE]+1) );

	printf("\t\tЗаполните таблицу \"Прокат\"\n\n");
	print_car_info();
	if( input_int(&rent_table[row].car_id, 10,
		"Введите код автомобиля (больше нуля) -> ") )
		return 1;
	if( rent_table[row].car_id <= 0 )
	{
		fprintf(stderr, "ОШИБКА: введен код меньше нуля, таблица не добавится");
		return 1;
	}

	print_client_info();
	if( input_int(&rent_table[row].client_id, 10,
		"Введите код клиента (больше нуля) -> ") )
		return 1;
	if( rent_table[row].client_id <= 0 )
	{
		fprintf(stderr, "ОШИБКА: введен код меньше нуля, таблица не добавится");
		return 1;
	}

	print_employee_info();
	if( input_int(&rent_table[row].employee_id, 10,
		"Введите код сотрудника (больше нуля) -> ") )
		return 1;
	if( rent_table[row].employee_id <= 0 )
	{
		fprintf(stderr, "ОШИБКА: введен код меньше нуля, таблица не добавится");
		return 1;
	}

	print_service_info();
	if( input_int(&rent_table[row].service1_id, 10,
		"Введите код 1-ой услуги (больше нуля) -> ") )
		return 1;
	if( rent_table[row].service1_id <= 0 )
	{
		fprintf(stderr, "ОШИБКА: введен код меньше нуля, таблица не добавится");
		return 1;
	}

	if( input_int(&rent_table[row].service2_id, 10,
		"Введите код 2-ой услуги (больше нуля) -> ") )
		return 1;
	if( rent_table[row].service2_id <= 0 )
	{
		fprintf(stderr, "ОШИБКА: введен код меньше нуля, таблица не добавится");
		return 1;
	}

	if( input_int(&rent_table[row].service3_id, 10,
		"Введите код 3-ой услуги (больше нуля) -> ") )
		return 1;
	if( rent_table[row].service2_id <= 0 )
	{
		fprintf(stderr, "ОШИБКА: введен код меньше нуля, таблица не добавится");
		return 1;
	}

	if( input_date(&rent_table[row].delivery_date, 10,
		"Введите дату выдачи") )
		return 1;
	if( rent_table[row].delivery_date.day < 0 ||
		rent_table[row].delivery_date.day > 31 ||
		rent_table[row].delivery_date.month < 0 ||
		rent_table[row].delivery_date.month > 12 ||
		rent_table[row].delivery_date.year < 0 )
		fprintf(stderr, "ПРЕДУПРЕЖДЕНИЕ: Скорее всего, введена не дата выдачи\n");

	if( input_int(&rent_table[row].rent_time, 10,
		"Введите срок проката -> ") )
		return 1;
	if( rent_table[row].rent_time < 0 )
		fprintf(stderr, "ПРЕДУПРЕЖДЕНИЕ: Скорее всего, введено не время проката\n");

	if( input_date(&rent_table[row].return_date, 10,
		"Введите дату возврата") )
		return 1;
	if( rent_table[row].return_date.day < 0 ||
		rent_table[row].return_date.day > 31 ||
		rent_table[row].return_date.month < 0 ||
		rent_table[row].return_date.month > 12 ||
		rent_table[row].return_date.year < 0 )
		fprintf(stderr, "ПРЕДУПРЕЖДЕНИЕ: Скорее всего, введена не дата возврата\n");

	if( input_int(&rent_table[row].rent_price, 10,
		"Введите цену проката -> ") )
		return 1;
	if( rent_table[row].rent_price < 0 )
		fprintf(stderr, "ПРЕДУПРЕЖДЕНИЕ: Скорее всего, введена не цена проката\n");

	if( input_string(rent_table[row].payment_mark, 255,
		"Введите пометку об оплате -> ") )
		return 1;

	table_row[RENT_TABLE]++;
	return 0;
}